File: neuralnetwork/Trainer.js
const {Network} = require("./Network")
const {GPU} = require("gpu.js")

class Trainer {
    model;
    generate;
    count;
    checks;
    mutation;
    answer;
    miss;
    generation = () =>{}

    constructor({model, generate, count, checks, mutation, answer, miss, generation}){
        this.model = model;
        this.generate = generate;
        this.count = count;
        this.checks = checks;
        this.mutation = mutation
        this.answer = checks
        this.answer = answer
        this.miss = miss
        this.generation = generation
        
    }

    train(generations) {
        var networks = []
        if(this.generate){
            for (var i = 0; i < this.count; i++) {
                networks[i] = new Network(this.model.inputs, this.model.outputs)
                networks[i].generate(this.model.depths)
            }
        }else{
            networks[0] = this.model;
            for (var i = 1; i < this.count; i++) {
                networks[i] = this.model.mutate(...this.mutation)
            }
        }


        for (var i = 0; i < generations; i++) {
            let closest, smallestOffset;
            for (var j = 0; j < networks.length; j++) {
                var offset = 0;
                for (var z = 0; z < this.checks; z++) {
                    let prompt = this.answer();
                    let a = networks[j].calculate(prompt.input)
                    offset += this.miss(a, prompt.output)
                }
                offset /= this.checks;
                if (!smallestOffset || offset < smallestOffset) {
                    smallestOffset = offset
                    closest = j;
                }
            }
            networks[0] = networks[closest];
            this.generation(i,networks[0],smallestOffset)
            for (var j = 1; j < networks.length; j++) {
                networks[j] = networks[closest].mutate(...this.mutation);
            }
        }
        return networks[0]

    }

    trainGPU(generations) {

        var networks = []
        if(this.generate){
            for (var i = 0; i < this.count; i++) {
                networks[i] = new Network(this.model.inputs, this.model.outputs)
                networks[i].generate(this.model.depths)
                networks[i] = networks[i].layers
            }
        }else{
            networks[0] = this.model;
            for (var i = 1; i < this.count; i++) {
                networks[i] = this.model.mutate(...this.mutation)
                networks[i] = networks[i].layers
            }
        }   
        const gpu = new GPU();
        const multiply = gpu.createKernel(function (a, b, length) {
            let sum = 0;
            for (let i = 0; i < length; i++) {
                sum += a[this.thread.z][this.thread.y][i] * b[this.thread.z][i][this.thread.x];
            }
            return sum;
        }, {
            dynamicOutput: true,
            dynamicArguments: true
        })
        
        for (let i = 0; i < generations; i++) {
            let input = []
            let output = []
            for (let j = 0; j < networks.length; j++) {
                input[j] = []
                output[j] = []
                for (let z = 0; z < this.checks; z++) {
                    let t = this.answer()
                    input[j][z] = t.input;
                    output[j][z] = t.output;
                }
            }
            
            for (let j = 0; j < networks[0].length; j++) {
                let layers = [];
                for(let z = 0; z < networks.length; z++){
                    layers[z] = networks[z][j]
                }
                multiply.setOutput([networks[0][j][0].length, this.checks, this.count])
                input = multiply(input,   layers  , networks[0][j].length);
            }
            
            let closest, smallestOffset;
            for (let j = 0; j < networks.length; j++) {
                let offset = 0;
                for (let z = 0; z < this.checks; z++) {
                    offset += this.miss(input[j][z], output[j][z])
                }
                offset /= this.checks
                if (!smallestOffset || offset < smallestOffset) {
                    smallestOffset = offset
                    closest = j;
                }
            }

            networks[0] = networks[closest]
            let t = new Network(this.model.inputs, this.model.outputs);
            t.layers = networks[0]
            this.generation(i,t,smallestOffset)
            for (let j = 1; j < networks.length; j++) {
                networks[j] = t.mutate(...this.mutation).layers
            }
        }

        let t = new Network(this.model.inputs, this.model.outputs);
        t.layers = networks[0]
        return t;

    }

}

module.exports = {Trainer}

File: cpp/NeuralNetwork/source/Trainer.cpp
#include "../headers/Trainer.h"


Trainer::Trainer(Network model_t, int count_t, int checks_t, float mutationPercentage_t, float mutationOffset_t, void (*answer_t)(double*, double*), double (*miss_t)(double*, double*), void (*generation_t)(int, Network, double)){
    model = model_t;
    count = count_t;
    checks = checks_t;
    mutationPercentage = mutationPercentage_t;
    mutationOffset = mutationOffset_t;
    answer = answer_t;
    miss = miss_t;
    generation = generation_t;
}

Network Trainer::train(int generations){
    Network* networks = (Network*)malloc(sizeof(Network) * count);
    networks[0] = model;
    int closest;
    double smallestOffset;
    double offset = 0;
    int cellCount = 0;
    int maxDepth = 0;

    for(int i = 1; i < count; i++){
        networks[i] = model.mutate(0.5, 0.1);
    }

    for(int i = 0; i < model.layerCount; i++){
        cellCount += model.depths[i] * model.depths[i+1];
        if(model.depths[i] > maxDepth){maxDepth = model.depths[i];}
    }

    if(model.depths[model.layerCount] > maxDepth){maxDepth = model.depths[model.layerCount];}

    double* inputLayer = (double*)malloc(sizeof(double) * maxDepth);
    double* outputLayer = (double*)malloc(sizeof(double) * maxDepth);
    double* expected = (double*)malloc(sizeof(double) * (model.depths[model.layerCount]));
    int layerOffset;
    double* t;

    for(int i = 0; i < generations; i++){

        for(int j = 0; j < count; j++){
            offset = 0;
            for(int z = 0; z < checks; z++){

                answer(inputLayer, expected);
                layerOffset = 0;
                for(int y = 0; y < model.layerCount; y++){
                    if(y != 0){
                        layerOffset += model.depths[y-1] * model.depths[y];
                    }
                    for(int c = 0; c < model.depths[y+1]; c++){
                        outputLayer[c] = 0;
                        for(int n = 0; n < model.depths[y]; n++){
                            outputLayer[c] += inputLayer[n] * networks[j].layers[layerOffset + n*model.depths[y+1] + c];
                        }
                        model.activation(outputLayer[c]);
                    }
                    t = inputLayer;
                    inputLayer = outputLayer;
                    outputLayer = t;
                }
                offset += miss(inputLayer, expected);
            
            }
            offset /= checks;
            if(j == 0 || offset < smallestOffset){
                smallestOffset = offset;
                closest = j;
            }
        }

        generation(i, networks[closest], smallestOffset);

        for(int j = 0; j < count; j++){
            if(j != closest){
                for(int z = 0; z < cellCount; z++){
                    if(float(rand() % 100) / 100 < mutationPercentage){
                        networks[j].layers[z] = networks[closest].layers[z] + (1 -float(rand() % 100) / 50) * mutationOffset;
                    }else{
                        networks[j].layers[z] = networks[closest].layers[z];

                    }
                }
            }
        }
        
    }

    free(inputLayer);
    free(outputLayer);
    free(expected);
    return networks[closest];

}

void threadFunc(Network* networks, int offset, int count, int checks, double& smallestOffset, int& index, void (*answer)(double*, double*), double (*miss)(double*, double*)){
    int closest;
    double smallestOffsetThread;
    int cellCount = 0;
    int maxDepth = 0;
    double inaccuracy;

    for(int i = 0; i < networks[0].layerCount; i++){
        cellCount += networks[0].depths[i] * networks[0].depths[i+1];
        if(networks[0].depths[i] > maxDepth){maxDepth = networks[0].depths[i];}
    }

    if(networks[0].depths[networks[0].layerCount] > maxDepth){maxDepth = networks[0].depths[networks[0].layerCount];}

    double* inputLayer = (double*)malloc(sizeof(double) * maxDepth);
    double* outputLayer = (double*)malloc(sizeof(double) * maxDepth);
    double* expected = (double*)malloc(sizeof(double) * (networks[0].depths[networks[0].layerCount]));
    int layerOffset;
    double* t;

    for(int j = offset; j < offset + count; j++){
            inaccuracy = 0;
            for(int z = 0; z < checks; z++){

                answer(inputLayer, expected);
                layerOffset = 0;
                for(int y = 0; y < networks[0].layerCount; y++){
                    if(y != 0){
                        layerOffset += networks[0].depths[y-1] * networks[0].depths[y];
                    }
                    for(int c = 0; c < networks[0].depths[y+1]; c++){
                        outputLayer[c] = 0;
                        for(int n = 0; n < networks[0].depths[y]; n++){
                            outputLayer[c] += inputLayer[n] * networks[j].layers[layerOffset + n*networks[0].depths[y+1] + c];
                        }
                        networks[0].activation(outputLayer[c]);
                    }
                    t = inputLayer;
                    inputLayer = outputLayer;
                    outputLayer = t;
                }
                inaccuracy += miss(inputLayer, expected);
            
            }
            inaccuracy /= checks;
            if(j == offset || inaccuracy < smallestOffsetThread){
                smallestOffsetThread = inaccuracy;
                closest = j;
            }
        }

        free(inputLayer);
        free(outputLayer);
        free(expected);
        if(index == -1 || smallestOffset > smallestOffsetThread){
            smallestOffset = smallestOffsetThread;
            index = closest;
        }
}


Network Trainer::train(int generations, int threads){
    Network* networks = (Network*)malloc(sizeof(Network) * count);
    networks[0] = model;
    std::thread workers[threads];
    int closest = -1;
    double smallestOffset;
    int cellCount = 0;
    int maxDepth = 0;
    int countPerThread = count / threads;

    for(int i = 1; i < count; i++){
        networks[i] = model.mutate(mutationPercentage, mutationOffset);
    }

    for(int i = 0; i < model.layerCount; i++){
        cellCount += model.depths[i] * model.depths[i+1];
        if(model.depths[i] > maxDepth){maxDepth = model.depths[i];}
    }

    if(model.depths[model.layerCount] > maxDepth){maxDepth = model.depths[model.layerCount];}

    for(int i = 0; i < generations; i++){

        int c = 0;
        closest = -1;
        for(int j = 0; j < threads; j++){
            if(j < count % threads){
                workers[j] = std::thread(threadFunc, std::ref(networks), j * countPerThread + c, countPerThread + 1, checks, std::ref(smallestOffset), std::ref(closest), answer, miss);
                c++;
            }else{
                workers[j] = std::thread(threadFunc, std::ref(networks), j * countPerThread + c, countPerThread , checks, std::ref(smallestOffset), std::ref(closest), std::ref(answer), std::ref(miss));
            }
        }

        for(int j = 0; j < threads; j++){
            workers[j].join();
        }

        generation(i , networks[closest], smallestOffset);

        for(int j = 0; j < count; j++){
            if(j != closest){
                for(int z = 0; z < cellCount; z++){
                    if(float(rand() % 100) / 100 < mutationPercentage){
                        networks[j].layers[z] = networks[closest].layers[z] + (1 -float(rand() % 100) / 50) * mutationOffset;
                    }else{
                        networks[j].layers[z] = networks[closest].layers[z];

                    }
                }
            }
        }
    }
    
    return networks[closest];

}

int Trainer::getComplexity(){
    return model.getComplexity()*count*checks;
}